File: 131_Palindrome Partitioning.java
// Time Complexity: O(n * 2^n)
// Space Complexity: O(n)

// Thought: similar as LC78, can be considered as a tree
class Solution {
    List<List<String>> res = new ArrayList<>();
    List<String> cur = new ArrayList<>();

    public List<List<String>> partition(String s) {
        bt(s, 0);
        return res;
    }

    private void bt(String s, int start) {
        // end case
        if (start == s.length()) {
            res.add(new ArrayList<>(cur));
            return;
        }

        // check all palindromes after the cur char
        for (int end = start; end < s.length(); end++) {
            if (checkPalindrome(s, start, end)) {
                cur.add(s.substring(start, end + 1));
                bt(s, end + 1);
                cur.remove(cur.size() - 1);
            }
        }
    }

    private boolean checkPalindrome(String s, int start, int end) {
        if (start == end) return true;

        while (start < end) {
            if (s.charAt(start++) != s.charAt(end--)) return false;
        }
        return true;
    }
}

File: 131_Palindrome Partitioning.py
class Solution:
    def partition(self, s: str) -> List[List[str]]:
        @lru_cache(None)
        def isPalindrom(left, right) -> bool:
            if left > right: return False
            while left < right:
                if s[left] != s[right]: return False
                left += 1
                right -= 1
            return True

        def dfs(start: int, idx: int):
            # end case
            if idx == len(s):
                if isPalindrom(start, idx-1):
                    curPath.append(s[start: idx])
                    res.append(list(curPath))
                    curPath.pop()
                return
            
            # split s[start:idx] or not
            if isPalindrom(start, idx-1):
                curPath.append(s[start: idx])
                dfs(idx, idx+1)
                curPath.pop()
            dfs(start, idx+1)

        curPath, res = [], []
        dfs(0, 0)
        return res